### ISSUE TITLE
spurious get() failure

### ISSUE THREAD

**Issue by rkuhn** [xacrimon/dashmap#139]
While testing some improvements for intern-arc using miri I found the following:

- one thread is doing `remove_if` for a key
- another thread for the same key sees `get` empty but `entry` still finds the entry that is being removed

This is probably a really weird scheduling issue, miri tends to surface those.

**Comment by xacrimon** [comment]
Which version are you on? The rc* versions were unstable test versions and shouldn't be used as they do not comply with semver.

**Comment by rkuhn** [comment]
This is on v3.11.4, see https://github.com/Actyx/intern-arc

**Comment by xacrimon** [comment]
Is entry() being run after get() and and finds the value while get doesn't?

**Comment by rkuhn** [comment]
Yes, at least that is the only possible explanation of the behaviour I observed.

**Comment by xacrimon** [comment]
Hmm, can you try on v4?

**Comment by rkuhn** [comment]
I have since changed my code, but I can try it in a few hours (first I must reproduce it with 3.11.4).

**Comment by rkuhn** [comment]
I pushed the result of updating intern-arc to 4.0.2: this has a different problem, surfacing in miri, where a race condition is detected between atomic store and dealloc in [this test case](https://github.com/Actyx/intern-arc/blob/9de43d6c3eef9aa6d7777c5d42523a7488cf9f1b/tests/smoke_and_miri.rs#L206-L232). The test below it passes, where the difference is only that instead of `DashMap` it uses `Mutex<BTreeMap<...>>` (which is more obviously correct).

In case you want to take a look:
```
error: Undefined Behavior: Data race detected between Deallocate on Thread(id = 1) and Atomic Store on Thread(id = 2), memory(alloc512364,offset=0,size=48)
(current vector clock = VClock([30, 37, 25]), conflicting timestamp = VClock([0, 27, 26]))
   --> /home/rkuhn/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/alloc.rs:104:14
    |
104 |     unsafe { __rust_dealloc(ptr, layout.size(), layout.align()) }
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Data race detected between Deallocate on Thread(id = 1) and Atomic Store on Thread(id = 2), memory(alloc512364,offset=0,size=48)
(current vector clock = VClock([30, 37, 25]), conflicting timestamp = VClock([0, 27, 26]))
    |
    = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
    = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
            
    = note: inside `std::alloc::dealloc` at /home/rkuhn/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/alloc.rs:104:14
    = note: inside `<intern_arc::Interned<TestStruct> as std::ops::Drop>::drop` at /home/rkuhn/comp/intern-arc/src/ref_count.rs:285:17
    = note: inside `std::intrinsics::drop_in_place::<intern_arc::Interned<TestStruct>> - shim(Some(intern_arc::Interned<TestStruct>))` at /home/rkuhn/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:187:1
    = note: inside `std::intrinsics::drop_in_place::<(intern_arc::Interned<TestStruct>, ())> - shim(Some((intern_arc::Interned<TestStruct>, ())))` at /home/rkuhn/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:187:1
    = note: inside `std::intrinsics::drop_in_place::<std::option::Option<(intern_arc::Interned<TestStruct>, ())>> - shim(Some(std::option::Option<(intern_arc::Interned<TestStruct>, ())>))` at /home/rkuhn/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:187:1
    = note: inside `intern_arc::hash::remover::<TestStruct>` at /home/rkuhn/comp/intern-arc/src/hash.rs:53:5
    = note: inside `<intern_arc::Interned<TestStruct> as std::ops::Drop>::drop` at /home/rkuhn/comp/intern-arc/src/ref_count.rs:250:17
    = note: inside `std::intrinsics::drop_in_place::<intern_arc::Interned<TestStruct>> - shim(Some(intern_arc::Interned<TestStruct>))` at /home/rkuhn/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:187:1
note: inside closure at tests/smoke_and_miri.rs:224:13
   --> tests/smoke_and_miri.rs:224:13
    |
224 |             }
    |             ^
[snip]
```

This means that one thread drops the interned value, sees ref_count=2, removes the entry from the DashMap [here](https://github.com/Actyx/intern-arc/blob/9de43d6c3eef9aa6d7777c5d42523a7488cf9f1b/src/hash.rs#L52). The other thread races to intern that value [here](https://github.com/Actyx/intern-arc/blob/9de43d6c3eef9aa6d7777c5d42523a7488cf9f1b/src/hash.rs#L144-L146), which should hold the read lock on the map (I also tried `get_mut`, the result is the same). So my analysis is:

- either the remove goes first, which means that interning should create a new RefCounted and not access the freed memory
- or the `get()` goes first, which includes the `entry.key().clone()`, so dropping the removed value [here](https://github.com/Actyx/intern-arc/blob/9de43d6c3eef9aa6d7777c5d42523a7488cf9f1b/src/hash.rs#L53) must again see ref_count=2 and not deallocate the memory

This leaves only the explanation that — at least for miri — the lock on the map does not actually guarantee mutual exclusion between `get` and `remove`. Or do you see another possibility? BTW, [here](https://github.com/Actyx/intern-arc/blob/9de43d6c3eef9aa6d7777c5d42523a7488cf9f1b/src/ref_count.rs#L312-L316)’s the Hash implementation for Interned.

**Comment by rkuhn** [comment]
BTW: why does DashMap use an RwLock that only spins? Wouldn’t parking_lot be more friendly in cases where Refs are kept around for more than a handful of cycles? (thinking priority inversion as well)

**Comment by xacrimon** [comment]
> BTW: why does DashMap use an RwLock that only spins? Wouldn’t parking_lot be more friendly in cases where Refs are kept around for more than a handful of cycles? (thinking priority inversion as well)

Absolutely, the reason for the current state is that parking_lot lacked a critical feature that was needed when the code was written.

**Comment by xacrimon** [comment]
Regarding your issue, dashmap itself is just an array of rwlocked maps and doesn't do much unsafe aside from skipping some array index checks. I think this is either an issue on your end or a MIRI bug. Sadly I lack the time to help you further. get() acquires a read lock on the shard for the key and remove acquires a write lock which does make them mutually exclusive.

**Comment by xacrimon** [comment]
Moving this to low-priority and needs-investigation. Significant changes has happened since now. If you can build a repro on v5 I'd be happy to assign this high-priority. Conversely, if it's gone, I can close this. Thank you.

**Comment by rkuhn** [comment]
Thanks for the update, I won’t be able to spend time on this in the near future, so feel free to close.

**Comment by xacrimon** [comment]
Reprod on 3.0 and seems to be fixed on 5.0. Closing.

**Comment by terorie** [comment]
Hi @xacrimon, would you mind sharing the code that reproduced the issue on 3.0? I'm curious as to whether it is possible to reproduce the same on 4.0.2. Thank you.
